File: Rosu/programmazione/laboratorio/java/progetti/huffman/src/ro/razvan/uniud/huffman/java/NodeQueue.java
package ro.razvan.uniud.huffman.java;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import static java.util.Arrays.deepToString;

public final class NodeQueue {

    @NotNull
    private Node[] queue;

    public NodeQueue() {
        queue = new Node[0];
    }

    /**
     * Aggiunge l'elemento alla queue mantenendo l'array ordinato
     *
     * @param n nuovo nodo da aggiungere alla queue
     */
    public void add(@NotNull final Node n) {

        final var size = size();

        var enough = false;

        final var newQueue = new Node[size + 1];

        for (int i = 0, j = 0; i < size; i++, j++) {

            if (queue[i].compareTo(n) > 0 && !enough) {
                newQueue[j++] = n;
                enough = true;
            }

            newQueue[j] = queue[i];

        }

        if (size == 0 || !enough) {
            newQueue[size] = n;
        }

        queue = newQueue;
    }

    /**
     * Rimuove dall'array l'elemento di minor peso e lo resituisce
     *
     * @return nodo di minor peso presente nell'array
     */
    @Nullable
    public Node poll() {

        final var size = size();

        if (size == 0) {
            return null;
        }

        final var element = queue[0];
        final var newQueue = new Node[size - 1];

        System.arraycopy(queue, 1, newQueue, 0, newQueue.length);

        queue = newQueue;

        return element;
    }

    /**
     * Trova la dimensione della queue
     *
     * @return dimensione della queue
     */
    public int size() {
        return queue.length;
    }

    @Override
    public String toString() {
        return deepToString(queue);
    }

}

File: Rosu/programmazione/laboratorio/java/progetti/huffman/src/ro/razvan/uniud/huffman/kotlin/NodeQueue.kt
package ro.razvan.uniud.huffman.kotlin

import ro.razvan.uniud.huffman.kotlin.util.deepToString

class NodeQueue {

    private var queue = arrayOfNulls<Node>(0)

    /**
     * Dimensione della queue
     */
    val size: Int
        get() = queue.size

    /**
     * Aggiunge l'elemento alla queue mantenendo l'array ordinato
     *
     * @param n nuovo nodo da aggiungere alla queue
     */
    fun add(n: Node) {

        var enough = false
        val queue = arrayOfNulls<Node>(size + 1)

        var j = 0

        for (i in 0 until this.queue.size) {

            if (!enough && this.queue[i]!! > n) {
                enough = true
                queue[j++] = n
            }

            queue[j++] = this.queue[i]

        }

        if (!enough || size == 0) {
            queue[size] = n
        }

        this.queue = queue

    }

    /**
     * Rimuove dall'array l'elemento di minor peso e lo resituisce
     *
     * @return nodo di minor peso presente nell'array
     */
    fun poll(): Node? {

        val element = queue.firstOrNull() ?: return null

        queue = queue.copyOfRange(1, size)

        return element
    }

    override fun toString(): String {
        return queue.deepToString()
    }


}